### PR TITLE
port feature "buffered columns"

### PR DESCRIPTION
**Pull request by Dronplane**

### CODE REVIEW

File: core/formats/columnstore.cpp [hunk 1]
```diff
@@ -2345,13 +2345,14 @@ class reader final : public columnstore_reader, public context_provider {
...
+               const options& opts = options{}) override;
 
-  virtual const column_reader* column(field_id field) const override;
+  const column_reader* column(field_id field) const override;
 
-  virtual bool visit(const column_visitor_f& visitor) const override;
+  bool visit(const column_visitor_f& visitor) const override;
 
-  virtual size_t size() const noexcept override { return columns_.size(); }
+  size_t size() const noexcept override { return columns_.size(); }
```
Comment (MBkkt) [inline core/formats/columnstore.cpp:2355]: final?

File: core/formats/columnstore2.hpp [hunk 2]
```diff
@@ -263,7 +264,8 @@ struct column_header {
 
 class reader final : public columnstore_reader {
  public:
-  virtual bool prepare(const directory& dir, const segment_meta& meta) override;
+  virtual bool prepare(const directory& dir, const segment_meta& meta,
+                       const options& opts = options{}) override;
```
Comment (MBkkt) [inline core/formats/columnstore2.hpp:268]:
```suggestion
  bool prepare(const directory& dir, const segment_meta& meta,
               const options& opts = options{}) final;
```

File: core/formats/formats.cpp [hunk 3]
```diff
@@ -76,6 +76,8 @@ class format_register
 
 namespace iresearch {
 
+bool noop_memory_accounter(int64_t) noexcept { return true; }
```
Comment (MBkkt) [inline core/formats/formats.cpp:79]:
```suggestion
constexpr bool noop_memory_accounter(int64_t) noexcept { return true; }
```

For example for trivial awaitable stdlib developers writiing something like this

Comment (MBkkt) [inline core/formats/formats.cpp:79]: What about PascalCase instead of snake_case?

File: core/index/segment_reader.hpp [hunk 4]
```diff
@@ -102,7 +103,8 @@ class segment_reader final : public sub_reader {
     return impl_->sort();
   }
 
-  virtual const irs::column_reader* column(std::string_view name) const override {
+  virtual const irs::column_reader* column(
+    std::string_view name) const override {
```
Comment (MBkkt) [inline core/index/segment_reader.hpp:107]:
```suggestion
  const irs::column_reader* column(
    std::string_view name) const final {
```

File: core/utils/file_utils.cpp [hunk 5]
```diff
@@ -751,6 +755,10 @@ handle_t open(const file_path_t path, OpenMode mode, int advice) noexcept {
       return handle_t(nullptr);
   }
   DWORD dwFlags = FILE_ATTRIBUTE_NORMAL | (static_cast<DWORD>(advice));
+  // Implement Alignment and File Access Requirements from MSDN
+  // if ((mode & OpenMode::Direct) == OpenMode::Direct) {
+  //  dwFlags |= FILE_FLAG_NO_BUFFERING;
```
Comment (MBkkt) [inline core/utils/file_utils.cpp:760]: Please create search task about o_direct reads

File: core/store/store_utils.hpp [hunk 6]
```diff
@@ -425,13 +425,56 @@ class bytes_view_input : public index_input {
...
+  virtual size_t read_bytes(size_t offset, byte_type* b,
+                            size_t size) noexcept override final {
+    return bytes_view_input::read_bytes(src_to_internal(offset), b, size);
+  }
+
+ private:
+  size_t src_to_internal(size_t t) const noexcept;
+
+  mapping mapping_;
+};  // remapped_bytes_view_input
```
Comment (MBkkt) [inline core/store/store_utils.hpp:476]:
```suggestion
};
```

File: core/store/store_utils.hpp [hunk 7]
```diff
@@ -425,13 +425,56 @@ class bytes_view_input : public index_input {
...
+  }
+
+  remapped_bytes_view_input(const remapped_bytes_view_input& other)
+    : bytes_view_input(other), mapping_{other.mapping_} {}
+
+  virtual int64_t checksum(size_t offset) const override final {
+    return bytes_view_input::checksum(src_to_internal(offset));
+  }
+
+  virtual void seek(size_t pos) noexcept override final {
```
Comment (MBkkt) [inline core/store/store_utils.hpp:457]:
```suggestion
  void seek(size_t pos) noexcept final {
```

File: core/store/store_utils.hpp [hunk 8]
```diff
@@ -425,13 +425,56 @@ class bytes_view_input : public index_input {
     return irs::vread<uint32_t>(pos_);
   }
 
-  virtual int64_t checksum(size_t offset) const override final;
+  virtual int64_t checksum(size_t offset) const override;
```
Comment (MBkkt) [inline core/store/store_utils.hpp:428]: override only

File: core/store/store_utils.cpp [hunk 9]
```diff
@@ -160,4 +160,38 @@ int64_t bytes_view_input::checksum(size_t offset) const {
...
 
+size_t remapped_bytes_view_input::src_to_internal(size_t t) const noexcept {
+  assert(!mapping_.empty());
+  auto it =
+    std::lower_bound(mapping_.begin(), mapping_.end(), t,
+                     [](const auto& l, const auto& r) { return l.first < r; });
+  if (it == mapping_.end()) {
+    --it;
+  } else if (it->first > t) {
+    assert(it != mapping_.begin());
```
Comment (MBkkt) [inline core/store/store_utils.cpp:171]: why is it true?

File: core/store/store_utils.cpp [hunk 10]
```diff
@@ -160,4 +160,38 @@ int64_t bytes_view_input::checksum(size_t offset) const {
...
+    --it;
+  }
+  return it->second + (t - it->first);
+}
+
+size_t remapped_bytes_view_input::file_pointer() const noexcept {
+  auto const adr = bytes_view_input::file_pointer();
+  size_t diff = std::numeric_limits<size_t>::max();
```
Comment (MBkkt) [inline core/store/store_utils.cpp:179]:
```suggestion
  auto diff = std::numeric_limits<size_t>::max();
```

File: core/store/store_utils.cpp [hunk 11]
```diff
@@ -160,4 +160,38 @@ int64_t bytes_view_input::checksum(size_t offset) const {
...
+    --it;
+  } else if (it->first > t) {
+    assert(it != mapping_.begin());
+    --it;
+  }
+  return it->second + (t - it->first);
+}
+
+size_t remapped_bytes_view_input::file_pointer() const noexcept {
+  auto const adr = bytes_view_input::file_pointer();
```
Comment (MBkkt) [inline core/store/store_utils.cpp:178]:
```suggestion
  auto const addr = bytes_view_input::file_pointer();
```

File: core/formats/columnstore2.cpp [hunk 12]
```diff
@@ -348,6 +348,17 @@ class column_base : public column_reader, private util::noncopyable {
...
+      auto released = static_cast<int64_t>(column_data_.size());
+      // force memory release
+      column_data_ = {};
+      if (buffered_memory_accounter_) {
+        buffered_memory_accounter_(-released);
+      }
+    }
+  }
+
   virtual std::string_view name() const final {
```
Comment (MBkkt) [inline core/formats/columnstore2.cpp:362]:
```suggestion
  std::string_view name() const final {
```

File: core/formats/columnstore2.cpp [hunk 13]
```diff
@@ -377,15 +388,81 @@ class column_base : public column_reader, private util::noncopyable {
...
+  bool allocate_buffered_memory(
+    size_t size, memory_accounting_f& buffered_memory_accounter) {
+    assert(size < static_cast<size_t>(std::numeric_limits<int64_t>::max()));
+    assert(buffered_memory_accounter);
+    if (!buffered_memory_accounter(static_cast<int64_t>(size))) {
+      // use string to have always null-terminated name
+      std::string col_name{"<anonymous>"};
+      if (!irs::IsNull(name())) {
+        col_name = name();
+      }
```
Comment (MBkkt) [inline core/formats/columnstore2.cpp:410]:
```suggestion
      const auto column_name = name();
      if (irs::IsNull(column_name)) {
        column_name = "<anonymous>";
      }
```

File: core/formats/columnstore2.cpp [hunk 14]
```diff
@@ -610,12 +687,41 @@ class dense_fixed_length_column final : public column_base {
 
   virtual doc_iterator::ptr iterator(ColumnHint hint) const override;
 
+  void make_buffered(
+    irs::index_input& in, memory_accounting_f& memory_accounter,
+    std::span<std::unique_ptr<column_reader>> next_sorted_columns) override {
```
Comment (MBkkt) [inline core/formats/columnstore2.cpp:692]:
```suggestion
    std::span<std::unique_ptr<column_reader>> next_sorted_columns) final {
```

**Review by MBkkt — APPROVED** [review]

File: core/formats/columnstore.cpp (hunk 1 above)
Comment (Dronplane) [reply core/formats/columnstore.cpp:2355]: Yeah. Good point.

File: core/store/store_utils.cpp (hunk 9 above)
Comment (Dronplane) [reply core/store/store_utils.cpp:171]: If it is false - remapped bytes input is used in a wrong way. Someone asked for offset that is before first mapping

Comment (Dronplane) [reply core/store/store_utils.cpp:171]: Should not happen. It is a bug  on call site if so

**Comment by Dronplane** [comment]
Tests blue except unrelated failures

**Pull request merged**